### ISSUE TITLE
FEATURE: Please allow both TCP 443 and UDP 443 on the server side.

### ISSUE THREAD

**Issue by sunny75016** [ultrahorizon/UH-VPN-Docs#27]
At the moment, for each server we have to explicitly choose UDP (on / off). For situations where the UDP traffic is blocked we are forced to use the TCP which reduces the efficiency for all users (including those who can use UDP). The alternative solution to have one server each for TCP and UDP is not acceptable as it simply multiplies the cost to Premium customers by two.

I would like UH-VPN to allow both TCP and UDP on the server side and the client side should first try to connect over UDP and switch over to TCP after a timeout. If you want to make this even more efficient the "Connect" UI can have two options instead i.e. Connect UDP / Connect TCP.

**Comment by jwsi** [comment]
Hi @sunny75016,

UH VPN servers are free for this reason. You can have as many servers as you wish in a premium group and it won't affect the price. Therefore, you should create a UH VPN UDP server and a UH VPN TCP server within the group and allow your clients to choose between the two within the client applications.

**Comment by sunny75016** [comment]
My apology jwsi - I should have explained better. The doubling of cost happens only because one has to create a new VPS (e.g. DigitalOcean droplet) one for TCP and one for UDP and incur a monthly cost for both servers. Perhaps I have missed a tick here.

**Comment by jwsi** [comment]
@sunny75016, sorry that makes sense. The good news is that each droplet/ubuntu instance can operate any number of UH VPN servers. To configure a droplet/ubuntu instance to operate more than one UH VPN server, simply add the token into the configuration file by following this guide:

https://docs.uh-vpn.com/en/latest/servers/adding-servers.html

Let me know if that solves things for you 👍

**Comment by sunny75016** [comment]
@jwsi - The problems is partly solved. Now the DigitalOcean Droplet VPS runs UH-VPN on both TCP/443 and UDP/443 and has accepted both tokens as per the logfile tail. My problems is that the connection on TCP/443 works perfectly well (iOS client) but not on UDP/443. I tried using WiFi as well as 4G - but simply unable to connect to the server. I checked that both tcp and udp are opened on ufw. I have attached the redacted-log file by removing token codes and IP addresss. Please help.
[redacted-log.txt](https://github.com/ultrahorizon/UH-VPN-Docs/files/4808943/redacted-log.txt)

**Comment by jwsi** [comment]
Hi @sunny75016 UH VPN server software definitely supports multiple servers, but here are some pitfalls that people often make:

1. Relevant protocol or port is **not open on a Digital Ocean firewall** attached to the droplet.
2. **Both UDP and TCP servers are using the same IP subnet**. UH VPN Servers running on the same droplet should always use independent tunnel networks for both IPv4 and IPv6 users. E.g. st UDP to `172.31.0.0/24` and TCP to `172.31.1.0/24`.

If you're still having issues after checking this, then let me know 👍

**Comment by sunny75016** [comment]
Very helpful, thanks @jwsi. I double checked firewall and ports 443/tcp 443/udp and 2802/tcp which were already open. Here are the results.

root@vpn:~# nc -z -v -u [IP redacted] 443
Connection to [IP redacted] port [udp/*] succeeded!
root@vpn:~# nc -z -v [IP redacted] 443
Connection to [IP redacted] port [tcp/https] succeeded!
root@vpn:~# nc -z -v [IP redacted] 2802
Connection to [IP redacted] port [tcp/*] succeeded!

In addition, the IPv4 and IPv6 subnets are as follows
TCP instance:  172.31.1.0/24  fe81::/64
UDP instance: 172.31.0.0/24  fe80::/64

I checked that the server has two separate tokens one for TCP and another for UDP.

After these changes to subnets, I have rebooted the VPS just in case. I also deleted the iOS VPN profile and created again (just in case). Still TCP/443 works fine but UDP/443 does not connect at all.  It is stuck at "Connecting to UH VPN..." both via WiFi and 4G.

Need your help (and I owe you a beer :-) )

**Comment by jwsi** [comment]
Hi @sunny75016, have you installed ufw onto the UH VPN Server droplet separately? If so can you try disabling that: 

`service ufw stop`
`sudo systemctl disable ufw`

Then reboot the VPS and test.
If it works, then we know this is the issue and it can probably be solved by turning off the "Auto add" rules in the server settings.

Let me know if turning off ufw temporarily solves the issue, once I know that then I'll be able to give advice on how we can get it working with ufw :)

**Comment by sunny75016** [comment]
Hi @jwsi - Thanks. I have now stopped and disabled ufw and rebooted VPS. I also ensured that the UDP/443 port is open and visible from the internet. Unfortunately VPS does not work on UDP at all; although it connects to TCP within milliseconds.  I deleted and recreated the whole server without firewall. Same issue.

**Comment by sunny75016** [comment]
I am using Ubuntu 20.04 - Has UH VPN been tested fully on it?

**Comment by jwsi** [comment]
Yes we have lots of people (including myself) who run UDP and TCP instances concurrently on the DO droplet.

Are you able to connect to UDP only server? This sounds like a censorship issue to me...

**Comment by jwsi** [comment]
In any case we'll get to the bottom of it, it'll probably be a simple config setting that we need to tweak, but I'm just trying to build a picture of what's going on at my end :)

**Comment by jwsi** [comment]
Hi @sunny75016, i just configured a VPS from DO with the parameters you have and have obtained the following configs and it works well. Can you please tell me if any of your outputs to the commands in my screenshots look different to yours:

<img width="926" alt="Service log" src="https://user-images.githubusercontent.com/16199862/85234249-ef3ee700-b403-11ea-9102-98863a83c3fa.png">
<img width="1295" alt="Log file 1" src="https://user-images.githubusercontent.com/16199862/85234251-f0701400-b403-11ea-9a67-4d90f87f6522.png">
<img width="1295" alt="Log file 2" src="https://user-images.githubusercontent.com/16199862/85234252-f239d780-b403-11ea-8879-54af5bc34bb4.png">
<img width="389" alt="forwarding" src="https://user-images.githubusercontent.com/16199862/85234253-f36b0480-b403-11ea-8f9c-41bed141480f.png">
<img width="382" alt="nat" src="https://user-images.githubusercontent.com/16199862/85234254-f4039b00-b403-11ea-8821-dd9e96ce418b.png">

**Comment by sunny75016** [comment]
Thanks @jwsi - I deleted the older droplet altogether and created a new Digital Ocean droplet. I followed the Ubuntu route for installation. I also deleted the older firewall altogether and created a new with the ports. Finally, I double checked that the subnets were different for TCP and UDP. End result - Both UDP and TCP installed on the same VPS are working perfectly fine using the iOS clients. 
I still owe you the beer....and I am glad we can treat this as permanently closed.
